File: software/examples/rust/example_callback.rs
use std::{error::Error, io, thread};
use tinkerforge::{accelerometer_bricklet::*, ip_connection::IpConnection};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Accelerometer Bricklet.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let a = AccelerometerBricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    // Create receiver for acceleration events.
    let acceleration_receiver = a.get_acceleration_receiver();

    // Spawn thread to handle received events. This thread ends when the `a` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for acceleration in acceleration_receiver {
            println!("Acceleration [X]: {} g", acceleration.x as f32 / 1000.0);
            println!("Acceleration [Y]: {} g", acceleration.y as f32 / 1000.0);
            println!("Acceleration [Z]: {} g", acceleration.z as f32 / 1000.0);
            println!();
        }
    });

    // Set period for acceleration receiver to 1s (1000ms).
    // Note: The acceleration callback is only called every second
    //       if the acceleration has changed since the last call!
    a.set_acceleration_callback_period(1000);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/go/example_callback.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/accelerometer_bricklet"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your Accelerometer Bricklet.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	a, _ := accelerometer_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	a.RegisterAccelerationCallback(func(x int16, y int16, z int16) {
		fmt.Printf("Acceleration [X]: %f g\n", float64(x)/1000.0)
		fmt.Printf("Acceleration [Y]: %f g\n", float64(y)/1000.0)
		fmt.Printf("Acceleration [Z]: %f g\n", float64(z)/1000.0)
		fmt.Println()
	})

	// Set period for acceleration receiver to 1s (1000ms).
	// Note: The acceleration callback is only called every second
	//       if the acceleration has changed since the last call!
	a.SetAccelerationCallbackPeriod(1000)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()
}
